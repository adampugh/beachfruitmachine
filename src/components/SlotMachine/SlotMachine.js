import React, { Component } from 'react';
import Modal from 'react-responsive-modal';

import Logo from '../../assets/logo.png';

class SlotMachine extends Component {
    state = {
        first: '',
        second: '',
        third: '',
        counter: 0,
        open: false,
        disabled: false
    }

    componentDidMount() {
        const { wheel } = this.props;
        this.setState({
            first: wheel[0],
            second: wheel[0],
            third: wheel[0]
        });
    }

    onOpenModal = () => {
        this.setState({ open: true });
    };
     
    onCloseModal = () => {
        this.setState({ open: false });
    };


    random = () => {
        const { wheel } = this.props
        return Math.floor(Math.random() * wheel.length);
    }


    handleSpin = () => {
        let { counter } = this.state;
        const { wheel } = this.props;

        const spin = setInterval(() => {

            // generate random numbers
            let left = this.random();
            let center = this.random();
            let right = this.random();

            // set state for current position
            this.setState({
                counter: counter++,
                first: wheel[left],
                second: wheel[center],
                third: wheel[right],
                disabled: true
            });

            // check result when finished spinning
            if (counter > 10) {
                this.setState({ 
                    counter: 0,
                    disabled: false
                });
                this.checkResult();
                clearInterval(spin);
            }
            
        }, 100);
    }
    
    checkResult = () => {
        const { first, second, third } = this.state; 

        if ((first === second) && (second === third)) {
            this.onOpenModal();
        }
    }

    render() {
        const { first, second, third, open, disabled } = this.state;

        return (
            <div className="slotMachine">
                <Modal open={open} onClose={this.onCloseModal} center>
                    <h1 className="slotMachine__modal__title">You Won!!!</h1>
                </Modal>
                <img className="slotMachine__logo" src={Logo} alt="logo" />
                <div className="slotMachine__grid">
                    <div className="slotMachine__slot">
                        <img src={first} alt="first slot" />
                    </div>
                    <div className="slotMachine__slot">
                        <img src={second} alt="second slot" />
                    </div>
                    <div className="slotMachine__slot">
                        <img src={third} alt="third slot" />
                    </div>
                </div>
                <div className="slotMachine__buttons">
                    <button className="btn" onClick={this.handleSpin} disabled={disabled}>Play</button>
                    <h3 className="slotMachine__buttons__text">Click the play button to start!</h3>
                </div>
            </div>
        )
    }
}

export default SlotMachine;