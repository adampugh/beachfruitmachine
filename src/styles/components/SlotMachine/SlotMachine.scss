.slotMachine {
    background: white;
    border-radius: 10px;
    box-sizing: border-box;
    margin: 40px 0; 
    text-align: center;
    @include boxshadow;
}


.slotMachine__grid {
    display: grid;
    grid-template-columns: 1fr 1fr 1fr;
    grid-gap: 10px;
    padding: 20px;
    img {
        max-width: 100%;
    }
}

.slotMachine__slot {
    align-self: center;
    border: 10px solid white;
    display: inline;
    @include boxshadow;
}

.slotMachine__buttons {
    text-align: center;
    padding-bottom: 40px;
}

.slotMachine__modal__title {
    padding: 100px;
    text-transform: uppercase;
    font-size: 2em;
}

.slotMachine__buttons__text {
    color: grey;
    padding-top: 20px;
}

.slotMachine__logo {
    max-width: 30%;
}