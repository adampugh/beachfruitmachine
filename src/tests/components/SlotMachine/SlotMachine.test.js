import React from 'react';
import '../../setupTests';
import { shallow } from 'enzyme';

import SlotMachine from '../../../components/SlotMachine/SlotMachine';

describe("<SlotMachine />", () => {
    it("should match snapshot", () => {
        const wheel = ['red', 'blue', 'green', 'yellow']
        const wrapper = shallow(<SlotMachine wheel={wheel}/>);
        expect(wrapper).toMatchSnapshot();
    });
});