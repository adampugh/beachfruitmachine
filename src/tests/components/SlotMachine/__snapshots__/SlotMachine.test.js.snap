// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SlotMachine /> should match snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SlotMachine
    wheel={
      Array [
        "red",
        "blue",
        "green",
        "yellow",
      ]
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Modal
          animationDuration={500}
          blockScroll={true}
          center={true}
          classNames={Object {}}
          classes={
            Object {
              "closeButton": "styles_closeButton__20ID4",
              "closeIcon": "styles_closeIcon__1QwbI",
              "modal": "styles_modal__gNwvD",
              "overlay": "styles_overlay__CLSq-",
              "overlayCenter": "styles_overlayCenter__YHoO7",
              "transitionEnter": "styles_transitionEnter__3j_-a",
              "transitionEnterActive": "styles_transitionEnterActive___eQs7",
              "transitionExit": "styles_transitionExit__1KmEf",
              "transitionExitActive": "styles_transitionExitActive__1nQXw",
            }
          }
          closeIconId={null}
          closeIconSize={28}
          closeIconSvgPath={
            <path
              d="M28.5 9.62L26.38 7.5 18 15.88 9.62 7.5 7.5 9.62 15.88 18 7.5 26.38l2.12 2.12L18 20.12l8.38 8.38 2.12-2.12L20.12 18z"
            />
          }
          closeOnEsc={true}
          closeOnOverlayClick={true}
          focusTrapOptions={Object {}}
          focusTrapped={false}
          modalId={null}
          onClose={[Function]}
          onEntered={null}
          onEscKeyDown={null}
          onExited={null}
          onOverlayClick={null}
          open={false}
          overlayId={null}
          showCloseIcon={true}
          styles={Object {}}
        >
          <h1
            className="slotMachine__modal__title"
          >
            You Won!!!
          </h1>
        </Modal>,
        <img
          alt="logo"
          className="slotMachine__logo"
          src="logo.png"
        />,
        <div
          className="slotMachine__grid"
        >
          <div
            className="slotMachine__slot"
          >
            <img
              alt="first slot"
              src="red"
            />
          </div>
          <div
            className="slotMachine__slot"
          >
            <img
              alt="second slot"
              src="red"
            />
          </div>
          <div
            className="slotMachine__slot"
          >
            <img
              alt="third slot"
              src="red"
            />
          </div>
        </div>,
        <div
          className="slotMachine__buttons"
        >
          <button
            className="btn"
            disabled={false}
            onClick={[Function]}
          >
            Play
          </button>
          <h3
            className="slotMachine__buttons__text"
          >
            Click the play button to start!
          </h3>
        </div>,
      ],
      "className": "slotMachine",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "animationDuration": 500,
          "blockScroll": true,
          "center": true,
          "children": <h1
            className="slotMachine__modal__title"
          >
            You Won!!!
          </h1>,
          "classNames": Object {},
          "classes": Object {
            "closeButton": "styles_closeButton__20ID4",
            "closeIcon": "styles_closeIcon__1QwbI",
            "modal": "styles_modal__gNwvD",
            "overlay": "styles_overlay__CLSq-",
            "overlayCenter": "styles_overlayCenter__YHoO7",
            "transitionEnter": "styles_transitionEnter__3j_-a",
            "transitionEnterActive": "styles_transitionEnterActive___eQs7",
            "transitionExit": "styles_transitionExit__1KmEf",
            "transitionExitActive": "styles_transitionExitActive__1nQXw",
          },
          "closeIconId": null,
          "closeIconSize": 28,
          "closeIconSvgPath": <path
            d="M28.5 9.62L26.38 7.5 18 15.88 9.62 7.5 7.5 9.62 15.88 18 7.5 26.38l2.12 2.12L18 20.12l8.38 8.38 2.12-2.12L20.12 18z"
          />,
          "closeOnEsc": true,
          "closeOnOverlayClick": true,
          "focusTrapOptions": Object {},
          "focusTrapped": false,
          "modalId": null,
          "onClose": [Function],
          "onEntered": null,
          "onEscKeyDown": null,
          "onExited": null,
          "onOverlayClick": null,
          "open": false,
          "overlayId": null,
          "showCloseIcon": true,
          "styles": Object {},
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "You Won!!!",
            "className": "slotMachine__modal__title",
          },
          "ref": null,
          "rendered": "You Won!!!",
          "type": "h1",
        },
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "alt": "logo",
          "className": "slotMachine__logo",
          "src": "logo.png",
        },
        "ref": null,
        "rendered": null,
        "type": "img",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="slotMachine__slot"
            >
              <img
                alt="first slot"
                src="red"
              />
            </div>,
            <div
              className="slotMachine__slot"
            >
              <img
                alt="second slot"
                src="red"
              />
            </div>,
            <div
              className="slotMachine__slot"
            >
              <img
                alt="third slot"
                src="red"
              />
            </div>,
          ],
          "className": "slotMachine__grid",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <img
                alt="first slot"
                src="red"
              />,
              "className": "slotMachine__slot",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "alt": "first slot",
                "src": "red",
              },
              "ref": null,
              "rendered": null,
              "type": "img",
            },
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <img
                alt="second slot"
                src="red"
              />,
              "className": "slotMachine__slot",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "alt": "second slot",
                "src": "red",
              },
              "ref": null,
              "rendered": null,
              "type": "img",
            },
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <img
                alt="third slot"
                src="red"
              />,
              "className": "slotMachine__slot",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "alt": "third slot",
                "src": "red",
              },
              "ref": null,
              "rendered": null,
              "type": "img",
            },
            "type": "div",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <button
              className="btn"
              disabled={false}
              onClick={[Function]}
            >
              Play
            </button>,
            <h3
              className="slotMachine__buttons__text"
            >
              Click the play button to start!
            </h3>,
          ],
          "className": "slotMachine__buttons",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Play",
              "className": "btn",
              "disabled": false,
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "Play",
            "type": "button",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Click the play button to start!",
              "className": "slotMachine__buttons__text",
            },
            "ref": null,
            "rendered": "Click the play button to start!",
            "type": "h3",
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Modal
            animationDuration={500}
            blockScroll={true}
            center={true}
            classNames={Object {}}
            classes={
              Object {
                "closeButton": "styles_closeButton__20ID4",
                "closeIcon": "styles_closeIcon__1QwbI",
                "modal": "styles_modal__gNwvD",
                "overlay": "styles_overlay__CLSq-",
                "overlayCenter": "styles_overlayCenter__YHoO7",
                "transitionEnter": "styles_transitionEnter__3j_-a",
                "transitionEnterActive": "styles_transitionEnterActive___eQs7",
                "transitionExit": "styles_transitionExit__1KmEf",
                "transitionExitActive": "styles_transitionExitActive__1nQXw",
              }
            }
            closeIconId={null}
            closeIconSize={28}
            closeIconSvgPath={
              <path
                d="M28.5 9.62L26.38 7.5 18 15.88 9.62 7.5 7.5 9.62 15.88 18 7.5 26.38l2.12 2.12L18 20.12l8.38 8.38 2.12-2.12L20.12 18z"
              />
            }
            closeOnEsc={true}
            closeOnOverlayClick={true}
            focusTrapOptions={Object {}}
            focusTrapped={false}
            modalId={null}
            onClose={[Function]}
            onEntered={null}
            onEscKeyDown={null}
            onExited={null}
            onOverlayClick={null}
            open={false}
            overlayId={null}
            showCloseIcon={true}
            styles={Object {}}
          >
            <h1
              className="slotMachine__modal__title"
            >
              You Won!!!
            </h1>
          </Modal>,
          <img
            alt="logo"
            className="slotMachine__logo"
            src="logo.png"
          />,
          <div
            className="slotMachine__grid"
          >
            <div
              className="slotMachine__slot"
            >
              <img
                alt="first slot"
                src="red"
              />
            </div>
            <div
              className="slotMachine__slot"
            >
              <img
                alt="second slot"
                src="red"
              />
            </div>
            <div
              className="slotMachine__slot"
            >
              <img
                alt="third slot"
                src="red"
              />
            </div>
          </div>,
          <div
            className="slotMachine__buttons"
          >
            <button
              className="btn"
              disabled={false}
              onClick={[Function]}
            >
              Play
            </button>
            <h3
              className="slotMachine__buttons__text"
            >
              Click the play button to start!
            </h3>
          </div>,
        ],
        "className": "slotMachine",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "animationDuration": 500,
            "blockScroll": true,
            "center": true,
            "children": <h1
              className="slotMachine__modal__title"
            >
              You Won!!!
            </h1>,
            "classNames": Object {},
            "classes": Object {
              "closeButton": "styles_closeButton__20ID4",
              "closeIcon": "styles_closeIcon__1QwbI",
              "modal": "styles_modal__gNwvD",
              "overlay": "styles_overlay__CLSq-",
              "overlayCenter": "styles_overlayCenter__YHoO7",
              "transitionEnter": "styles_transitionEnter__3j_-a",
              "transitionEnterActive": "styles_transitionEnterActive___eQs7",
              "transitionExit": "styles_transitionExit__1KmEf",
              "transitionExitActive": "styles_transitionExitActive__1nQXw",
            },
            "closeIconId": null,
            "closeIconSize": 28,
            "closeIconSvgPath": <path
              d="M28.5 9.62L26.38 7.5 18 15.88 9.62 7.5 7.5 9.62 15.88 18 7.5 26.38l2.12 2.12L18 20.12l8.38 8.38 2.12-2.12L20.12 18z"
            />,
            "closeOnEsc": true,
            "closeOnOverlayClick": true,
            "focusTrapOptions": Object {},
            "focusTrapped": false,
            "modalId": null,
            "onClose": [Function],
            "onEntered": null,
            "onEscKeyDown": null,
            "onExited": null,
            "onOverlayClick": null,
            "open": false,
            "overlayId": null,
            "showCloseIcon": true,
            "styles": Object {},
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "You Won!!!",
              "className": "slotMachine__modal__title",
            },
            "ref": null,
            "rendered": "You Won!!!",
            "type": "h1",
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "alt": "logo",
            "className": "slotMachine__logo",
            "src": "logo.png",
          },
          "ref": null,
          "rendered": null,
          "type": "img",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div
                className="slotMachine__slot"
              >
                <img
                  alt="first slot"
                  src="red"
                />
              </div>,
              <div
                className="slotMachine__slot"
              >
                <img
                  alt="second slot"
                  src="red"
                />
              </div>,
              <div
                className="slotMachine__slot"
              >
                <img
                  alt="third slot"
                  src="red"
                />
              </div>,
            ],
            "className": "slotMachine__grid",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <img
                  alt="first slot"
                  src="red"
                />,
                "className": "slotMachine__slot",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "alt": "first slot",
                  "src": "red",
                },
                "ref": null,
                "rendered": null,
                "type": "img",
              },
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <img
                  alt="second slot"
                  src="red"
                />,
                "className": "slotMachine__slot",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "alt": "second slot",
                  "src": "red",
                },
                "ref": null,
                "rendered": null,
                "type": "img",
              },
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <img
                  alt="third slot"
                  src="red"
                />,
                "className": "slotMachine__slot",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "alt": "third slot",
                  "src": "red",
                },
                "ref": null,
                "rendered": null,
                "type": "img",
              },
              "type": "div",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <button
                className="btn"
                disabled={false}
                onClick={[Function]}
              >
                Play
              </button>,
              <h3
                className="slotMachine__buttons__text"
              >
                Click the play button to start!
              </h3>,
            ],
            "className": "slotMachine__buttons",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Play",
                "className": "btn",
                "disabled": false,
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "Play",
              "type": "button",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Click the play button to start!",
                "className": "slotMachine__buttons__text",
              },
              "ref": null,
              "rendered": "Click the play button to start!",
              "type": "h3",
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
